crates/swc_atoms/src/wtf8_atom.rs

```diff
@@ -0,0 +1,237 @@
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Debug for Wtf8Atom {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
+    }
+}
+
+impl PartialOrd for Wtf8Atom {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Wtf8Atom {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_wtf8().cmp(other.as_wtf8())
+    }
+}
+
+impl<T> PartialEq<T> for Wtf8Atom
+where
+    hstr::Wtf8Atom: PartialEq<T>,
+    T: ?Sized,
+{
+    fn eq(&self, other: &T) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl<T> From<T> for Wtf8Atom
+where
+    hstr::Wtf8Atom: From<T>,
+{
+    fn from(s: T) -> Self {
+        Wtf8Atom::new(s)
+    }
+}
+
+impl From<&Wtf8Atom> for Wtf8Buf {
+    fn from(s: &Wtf8Atom) -> Self {
+        // SAFETY: `Wtf8Atom` is guaranteed to be valid WTF-8 byte sequence.
+        unsafe { Wtf8Buf::from_bytes_unchecked(s.as_bytes().to_vec()) }
+    }
+}
+
+impl serde::ser::Serialize for Wtf8Atom {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for Wtf8Atom {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        hstr::Wtf8Atom::deserialize(deserializer).map(Wtf8Atom)
+    }
+}
+
+/// NOT A PUBLIC API
+#[cfg(feature = "rkyv-impl")]
+impl rkyv::Archive for Wtf8Atom {
+    type Archived = rkyv::vec::ArchivedVec<u8>;
+    type Resolver = rkyv::vec::VecResolver;
+
+    #[allow(clippy::unit_arg)]
+    fn resolve(&self, resolver: Self::Resolver, out: rkyv::Place<Self::Archived>) {
+        rkyv::vec::ArchivedVec::<u8>::resolve_from_slice(self.as_bytes(), resolver, out)
+    }
+}
+
+/// NOT A PUBLIC API
+#[cfg(feature = "rkyv-impl")]
+impl<S: rancor::Fallible + rkyv::ser::Writer + rkyv::ser::Allocator + ?Sized> rkyv::Serialize<S>
+    for Wtf8Atom
+where
+    <S as rancor::Fallible>::Error: rancor::Source,
+{
+    fn serialize(&self, serializer: &mut S) -> Result<Self::Resolver, S::Error> {
+        rkyv::vec::ArchivedVec::<u8>::serialize_from_slice(self.as_bytes(), serializer)
+    }
+}
+
+/// NOT A PUBLIC API
+#[cfg(feature = "rkyv-impl")]
+impl<D> rkyv::Deserialize<Wtf8Atom, D> for rkyv::vec::ArchivedVec<u8>
+where
+    D: ?Sized + rancor::Fallible,
+    <D as rancor::Fallible>::Error: rancor::Source,
+{
+    fn deserialize(
+        &self,
+        deserializer: &mut D,
+    ) -> Result<Wtf8Atom, <D as rancor::Fallible>::Error> {
+        let s: Vec<u8> = self.deserialize(deserializer)?;
+
+        Ok(Wtf8Atom::new(
+            // SAFETY: `ArchivedVec<u8>` is guaranteed to be serialized with `Wtf8Atom` byte
+            // sequence.  `Wtf8Atom` byte sequence is identical to `Wtf8` byte sequence.
+            unsafe { Wtf8::from_bytes_unchecked(&s) },
```

Comment:
> `ArchivedVec<u8>` is guaranteed to be serialized with `Wtf8Atom` byte sequence.

This guarantee doesn't seem to be enforced, and users can construct a non-wtf8 `ArchivedVec<u8>`?

Reply:
Would you please shed more light on the scenario that you're referring to?

Previously, I was thinking about the case of mixing old plugins with a newer core:

- a new `Str` core + old plugins: will deserialize it from `ArchivedString`, I assume it will return an `Result::Err` as the `ArchivedVec<u8>` is different from `ArchivedString`
- a new `Str` core + new plugins , will deserialize it from `ArchivedVec<u8>` and it will be a success
- an old `Str` core + new plugins, will deserialize it with `ArchivedVec<u8>`, and it will fail as the original value is an `ArchivedString`
- an old `Str` core + old plugins, will be fine.